architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+UCP is a multi-location control-plane like the control-plane that inspired it (ARM). We don't use UCP/Radius in any multi-location scenarios right now but it's important we don't design APIs that lock us into a single-region design. ARM has a mature multi-location design but doesn't expose to users the ability to register types. However we can learn a lot from the scenarios supported by ARM for rolling out new API versions and new types.
+
+- Resource providers (non-declarative) are deployed per-location.
+- New functionality (API versions + types) is rolled out per-location.
+
+Some examples:
+
+> Resource provider `Contoso.Example` is only available in the `westus-1` location. UCP must reject creation requests to other locations.
+
+> Resource type `Contoso.Example/widgets` is available in all locations, but new API version `2024-08-01-preview` is only in the `eastus-2` location. UCP should block all requests for unsupported API version x location pairs.
+
+The design proposal includes axes for configuring location affinity:
+
+- `System.Resources/resourceProviders` includes a list of supported locations with an optional `address` in each location for the resource provider.
+- `System.Resources/resourceProviders/resourceTypes` includes a list of locations where the resource type is supported and the API versions supported in each of those locations.
+
+**Open Question:** There's a few possible designs for this that might make sense. Right now location shows up as a concept in both `resourceProviders` and `resourceTypes`. It would be possible to move all of this to `resourceProviders`. It would also be possible to build another resource type `System.Resources/resourceProviders/locations` for the location configuration. Let's discuss.
+
+### Architecture Diagram
+
+Resource type registration and querying is implemented in UCP. Any client may call the registration APIs or query APIs including resource providers.
+
+<center>
+
+![architecture diagram](./2024-08-user-defined-types-registration/registration.png)
+
+**Fig: Diagram of resource type registration and querying** [link](https://excalidraw.com/#json=FzQQaxv1d-vlmmDBAeOfc,UUXIasjO8p27hN0AB8qeiw)
+
+</center>
+
+### Detailed Design
+
+Since this is our first nested resource type, a basic primer on nested types is needed. A nested resource type exists *inside* another resource type. A resource type is that is not nested is *top-level*. These relationships are also sometimes referred to as a parent and child. 
+
+The following statements are all accurate:
+
+- The `System.Resources/resourceProviders` type is the parent of `System.Resources/resourceProviders/resourceTypes`. 
+- The `System.Resources/resourceProviders/resourceTypes` type is the child of `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders/resourceTypes` type is nested within `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders` type is a top-level resource type.
```

Comment:
We can't have a resource type being a parent of another resource type, right? Like going deeper in different types of **redis** for example.

Reply:
That's a great question and something we need to support. Let me think about how we're going to represent that.